File: app/agents/streaming/event_types.py
"""Typed event dataclasses for AG-UI streaming events.

These types define the contract between the backend agent and frontend UI
for all custom events emitted during agent execution.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Literal, Optional


@dataclass
class TraceStep:
    """A single step in the agent's thinking trace.

    Represents either a thought (model reasoning), action (tool call),
    or result (tool output or final response).
    """

    id: str
    timestamp: str
    type: Literal["thought", "action", "result"]
    content: str
    metadata: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def create(
        cls,
        step_id: str,
        step_type: Literal["thought", "action", "result"],
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> "TraceStep":
        """Factory method to create a TraceStep with current timestamp."""
        return cls(
            id=step_id,
            timestamp=datetime.now(timezone.utc).isoformat(),
            type=step_type,
            content=content,
            metadata=metadata or {},
        )

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary."""
        return {
            "id": self.id,
            "timestamp": self.timestamp,
            "type": self.type,
            "content": self.content,
            "metadata": _safe_metadata(self.metadata),
        }


@dataclass
class TimelineOperation:
    """An operation in the agent timeline.

    Operations form a tree structure representing the agent's execution:
    - Root: The main agent execution
    - Children: Tool calls, thoughts, and todos
    """

    id: str
    type: Literal["agent", "tool", "thought", "todo"]
    name: str
    status: Literal["pending", "running", "success", "error"]
    parent: Optional[str] = None
    children: List[str] = field(default_factory=list)
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    duration: Optional[int] = None  # milliseconds
    metadata: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def create_root(cls, root_id: str, name: str = "Unified Agent", **metadata: Any) -> "TimelineOperation":
        """Create the root agent operation."""
        return cls(
            id=root_id,
            type="agent",
            name=name,
            status="running",
            start_time=datetime.now(timezone.utc).isoformat(),
            metadata=metadata,
        )

    @classmethod
    def create_tool(
        cls,
        tool_call_id: str,
        tool_name: str,
        parent_id: str,
        **metadata: Any,
    ) -> "TimelineOperation":
        """Create a tool operation."""
        return cls(
            id=tool_call_id,
            type="tool",
            name=tool_name,
            status="running",
            parent=parent_id,
            start_time=datetime.now(timezone.utc).isoformat(),
            metadata={"toolName": tool_name, **metadata},
        )

    @classmethod
    def create_thought(
        cls,
        run_id: str,
        parent_id: str,
        model: Optional[str] = None,
    ) -> "TimelineOperation":
        """Create a thought/reasoning operation."""
        return cls(
            id=run_id,
            type="thought",
            name="Thinking",
            status="running",
            parent=parent_id,
            start_time=datetime.now(timezone.utc).isoformat(),
            metadata={"model": model} if model else {},
        )

    @classmethod
    def create_todo(
        cls,
        todo_id: str,
        title: str,
        parent_id: str,
        status: str = "pending",
        todo_data: Optional[Dict[str, Any]] = None,
    ) -> "TimelineOperation":
        """Create a todo operation."""
        op_status: Literal["pending", "running", "success", "error"] = "pending"
        if status == "in_progress":
            op_status = "running"
        elif status == "done":
            op_status = "success"

        return cls(
            id=todo_id,
            type="todo",
            name=title,
            status=op_status,
            parent=parent_id,
            metadata={"todo": todo_data or {}},
        )

    def complete(self, success: bool = True) -> None:
        """Mark operation as completed and calculate duration."""
        now = datetime.now(timezone.utc)
        self.end_time = now.isoformat()
        self.status = "success" if success else "error"

        if self.start_time:
            try:
                start_dt = datetime.fromisoformat(self.start_time)
                self.duration = int((now - start_dt).total_seconds() * 1000)
            except Exception:
                pass

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary."""
        result: Dict[str, Any] = {
            "id": self.id,
            "type": self.type,
            "name": self.name,
            "status": self.status,
            "children": list(self.children),
        }
        if self.parent:
            result["parent"] = self.parent
        if self.start_time:
            result["startTime"] = self.start_time
        if self.end_time:
            result["endTime"] = self.end_time
        if self.duration is not None:
            result["duration"] = self.duration
        if self.metadata:
            result["metadata"] = _safe_metadata(self.metadata)
        return result


@dataclass
class TodoItem:
    """A single todo item."""

    id: str
    title: str
    status: Literal["pending", "in_progress", "done"]
    metadata: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary."""
        return {
            "id": self.id,
            "title": self.title,
            "status": self.status,
            "metadata": _safe_metadata(self.metadata),
        }


@dataclass
class AgentThinkingTraceEvent:
    """Event payload for agent_thinking_trace custom event.

    Sent to update the thinking trace sidebar with model reasoning steps.
    """

    total_steps: int
    thinking_trace: Optional[List[TraceStep]] = None
    latest_step: Optional[TraceStep] = None
    active_step_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary with camelCase keys."""
        result: Dict[str, Any] = {"totalSteps": self.total_steps}

        if self.thinking_trace is not None:
            result["thinkingTrace"] = [step.to_dict() for step in self.thinking_trace]
        if self.latest_step is not None:
            result["latestStep"] = self.latest_step.to_dict()
        if self.active_step_id is not None:
            result["activeStepId"] = self.active_step_id

        return result


@dataclass
class AgentTimelineUpdateEvent:
    """Event payload for agent_timeline_update custom event.

    Sent to update the agentic timeline visualization.
    """

    operations: List[TimelineOperation]
    current_operation_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary with camelCase keys."""
        result: Dict[str, Any] = {
            "operations": [op.to_dict() for op in self.operations],
        }
        if self.current_operation_id is not None:
            result["currentOperationId"] = self.current_operation_id
        return result


@dataclass
class ToolEvidenceUpdateEvent:
    """Event payload for tool_evidence_update custom event.

    Sent to surface tool output and evidence in the Tool Evidence sidebar.
    """

    tool_call_id: str
    tool_name: str
    output: Any
    summary: Optional[str] = None
    cards: List[Dict[str, Any]] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary with camelCase keys."""
        result: Dict[str, Any] = {
            "toolCallId": self.tool_call_id,
            "toolName": self.tool_name,
            "output": _safe_json_value(self.output),
            "cards": self.cards,
        }
        if self.summary is not None:
            result["summary"] = self.summary
        return result


@dataclass
class AgentTodosUpdateEvent:
    """Event payload for agent_todos_update custom event.

    Sent to update the task/todo list in the Run Tasks sidebar.
    """

    todos: List[TodoItem]

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to JSON-safe dictionary."""
        return {"todos": [todo.to_dict() for todo in self.todos]}


# Utility functions

def _safe_metadata(metadata: Dict[str, Any]) -> Dict[str, Any]:
    """Ensure all metadata values are JSON-serializable."""
    import json

    safe: Dict[str, Any] = {}
    for key, value in metadata.items():
        try:
            json.dumps(value)
            safe[key] = value
        except TypeError:
            safe[key] = str(value)
    return safe


def _safe_json_value(value: Any) -> Any:
    """Convert a value to be JSON-serializable."""
    import json

    try:
        json.dumps(value, ensure_ascii=False)
        return value
    except TypeError:
        if isinstance(value, str):
            trimmed = value.strip()
            if (trimmed.startswith("{") and trimmed.endswith("}")) or (
                trimmed.startswith("[") and trimmed.endswith("]")
            ):
                try:
                    parsed = json.loads(trimmed)
                    json.dumps(parsed, ensure_ascii=False)
                    return parsed
                except Exception:
                    return value
            return value
        return str(value)

File: frontend/src/services/ag-ui/event-types.ts
/**
 * AG-UI Event Types
 *
 * TypeScript types mirroring backend event_types.py exactly.
 * These types define the contract for custom events emitted by Agent Sparrow.
 */

// -----------------------------------------------------------------------------
// Thinking Trace Types
// -----------------------------------------------------------------------------

export type TraceStepType = "thought" | "action" | "result" | "tool";

/**
 * A single step in the agent's thinking trace.
 * Emitted via `agent_thinking_trace` custom events.
 */
export interface TraceStep {
  /** Unique identifier for this step */
  id: string;
  /** ISO timestamp of when this step occurred */
  timestamp: string;
  /** Type of step */
  type: TraceStepType;
  /** Human-readable content describing this step */
  content: string;
  /** Additional metadata (tool info, model info, etc.) */
  metadata: Record<string, unknown>;
  /** Optional tool name if type is tool/action */
  toolName?: string;
  /** Optional duration in seconds */
  duration?: number;
  /** Optional status of the step */
  status?: "success" | "error";
}

/**
 * Event payload for `agent_thinking_trace` custom events.
 */
export interface AgentThinkingTraceEvent {
  /** Total number of steps in the trace so far */
  totalSteps: number;
  /** Full trace if syncing all steps (usually on first event) */
  thinkingTrace?: TraceStep[];
  /** Latest step if incremental update */
  latestStep?: TraceStep;
  /** ID of the currently active step */
  activeStepId?: string;
}

// -----------------------------------------------------------------------------
// Timeline Operation Types
// -----------------------------------------------------------------------------

export type TimelineOperationType = "agent" | "tool" | "thought" | "todo";
export type TimelineOperationStatus =
  | "pending"
  | "running"
  | "success"
  | "error";

/**
 * A single operation in the agent timeline.
 * Represents agent runs, tool calls, thinking, and todo items.
 */
export interface TimelineOperation {
  /** Unique identifier for this operation */
  id: string;
  /** Type of operation */
  type: TimelineOperationType;
  /** Human-readable name */
  name: string;
  /** Current status */
  status: TimelineOperationStatus;
  /** Parent operation ID (for nesting) */
  parent?: string;
  /** Child operation IDs */
  children: string[];
  /** ISO timestamp when operation started */
  startTime?: string;
  /** ISO timestamp when operation ended */
  endTime?: string;
  /** Duration in milliseconds */
  duration?: number;
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Event payload for `agent_timeline_update` custom events.
 */
export interface AgentTimelineUpdateEvent {
  /** All operations in the timeline */
  operations: TimelineOperation[];
  /** ID of the currently active operation */
  currentOperationId?: string;
}

// -----------------------------------------------------------------------------
// Tool Evidence Types
// -----------------------------------------------------------------------------

export interface ToolEvidenceCard {
  id?: string;
  type?: string;
  title?: string;
  snippet?: string;
  url?: string;
  fullContent?: unknown;
  status?: string;
  timestamp?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Event payload for `tool_evidence_update` custom events.
 * Provides detailed tool output for display in the reasoning panel.
 */
export interface ToolEvidenceUpdateEvent {
  /** Tool call identifier (matches tool_calls in AIMessage) */
  toolCallId: string;
  /** Name of the tool */
  toolName: string;
  /** Raw output from the tool */
  output: unknown;
  /** Alternative field names for output (backend compatibility) */
  result?: unknown;
  data?: unknown;
  value?: unknown;
  /** Tool input arguments */
  args?: string | Record<string, unknown>;
  /** Human-readable summary of the output */
  summary?: string;
  /** Pre-built evidence cards from the backend (optional) */
  cards?: ToolEvidenceCard[];
  /** Additional metadata from the tool call (optional) */
  metadata?: Record<string, unknown>;
}

// -----------------------------------------------------------------------------
// Todo Types
// -----------------------------------------------------------------------------

export type TodoStatus = "pending" | "in_progress" | "done";

/**
 * A single todo item created by the agent.
 */
export interface TodoItem {
  /** Unique identifier */
  id: string;
  /** Todo title/description */
  title: string;
  /** Current status */
  status: TodoStatus;
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Event payload for `agent_todos_update` custom events.
 */
export interface AgentTodosUpdateEvent {
  /** All current todo items */
  todos: TodoItem[];
}

// -----------------------------------------------------------------------------
// GenUI State Types
// -----------------------------------------------------------------------------

/**
 * Event payload for `genui_state_update` custom events.
 * Contains arbitrary state for generative UI components.
 */
export interface GenuiStateUpdateEvent {
  [key: string]: unknown;
}

/**
 * Event payload for `image_artifact` custom events.
 * Contains image data for frontend artifact display.
 */
export interface ImageArtifactEvent {
  id: string;
  type: "image";
  title: string;
  content: string;
  messageId: string;
  /** Preferred (Phase V): retrievable image URL (no base64 payloads). */
  imageUrl?: string;
  /** Legacy (pre-Phase V): base64 image payload. */
  imageData?: string;
  mimeType: string;
  altText?: string;
  aspectRatio?: string;
  resolution?: string;
  /** Optional source page URL for web-sourced images. */
  pageUrl?: string;
}

/**
 * Event payload for `article_artifact` custom events.
 * Contains article markdown for frontend artifact display.
 */
export interface ArticleArtifactEvent {
  id: string;
  type: "article";
  title: string;
  content: string;
  messageId: string;
  images?: Array<{
    url?: string;
    alt?: string;
    pageUrl?: string;
    page_url?: string;
  }>;
}

// -----------------------------------------------------------------------------
// Subagent Types
// -----------------------------------------------------------------------------

export interface SubagentSpawnEvent {
  subagentType: string;
  toolCallId: string;
  parentAgentId?: string;
  task: string;
  timestamp: string;
}

export interface SubagentEndEvent {
  subagentType: string;
  toolCallId: string;
  status: "success" | "error";
  reportPath: string;
  excerpt: string;
  timestamp: string;
}

export interface SubagentThinkingDeltaEvent {
  toolCallId: string;
  delta: string;
  timestamp: string;
  subagentType?: string;
}

// -----------------------------------------------------------------------------
// Union Type for All Custom Events
// -----------------------------------------------------------------------------

/**
 * Union type representing all possible custom event payloads.
 * Use this for type-safe event handling.
 */
export type AgentCustomEvent =
  | { name: "agent_thinking_trace"; value: AgentThinkingTraceEvent }
  | { name: "agent_timeline_update"; value: AgentTimelineUpdateEvent }
  | { name: "tool_evidence_update"; value: ToolEvidenceUpdateEvent }
  | { name: "agent_todos_update"; value: AgentTodosUpdateEvent }
  | { name: "genui_state_update"; value: GenuiStateUpdateEvent }
  | { name: "image_artifact"; value: ImageArtifactEvent }
  | { name: "article_artifact"; value: ArticleArtifactEvent }
  | { name: "subagent_spawn"; value: SubagentSpawnEvent }
  | { name: "subagent_end"; value: SubagentEndEvent }
  | { name: "subagent_thinking_delta"; value: SubagentThinkingDeltaEvent };

/**
 * Known AG-UI custom event names as const tuple for compile-time type derivation.
 */
export const KNOWN_EVENT_NAMES = [
  "agent_thinking_trace",
  "agent_timeline_update",
  "tool_evidence_update",
  "agent_todos_update",
  "genui_state_update",
  "image_artifact",
  "article_artifact",
  "subagent_spawn",
  "subagent_end",
  "subagent_thinking_delta",
] as const;

/**
 * Type derived from KNOWN_EVENT_NAMES for compile-time safety.
 */
export type KnownEventName = (typeof KNOWN_EVENT_NAMES)[number];

/**
 * Type guard to check if an event is a known AG-UI custom event.
 *
 * Performs shallow validation of the event structure:
 * - Checks that event is a non-null object
 * - Validates that event.name is a known event name string
 * - Validates that event.value exists and is an object
 *
 * @remarks For full validation of event.value shape, use the Zod validators
 * from validators.ts after this guard passes.
 */
export function isAgentCustomEvent(event: unknown): event is AgentCustomEvent {
  if (!event || typeof event !== "object") return false;
  const e = event as Record<string, unknown>;

  // Validate event.name is a known event name
  if (typeof e.name !== "string") return false;
  if (!(KNOWN_EVENT_NAMES as readonly string[]).includes(e.name)) return false;

  // Shallow validation: event.value must be present and be an object
  // Full validation of value shape should be done via Zod validators
  if (!("value" in e) || e.value === null || typeof e.value !== "object") {
    return false;
  }

  return true;
}

/**
 * Extract the event name from a custom event.
 */
export function getEventName(
  event: AgentCustomEvent,
): AgentCustomEvent["name"] {
  return event.name;
}

// -----------------------------------------------------------------------------
// Type Utilities
// -----------------------------------------------------------------------------

/**
 * Extract the value type for a specific event name.
 */
export type EventValueType<T extends AgentCustomEvent["name"]> = Extract<
  AgentCustomEvent,
  { name: T }
>["value"];

/**
 * Helper to create a typed event payload.
 */
export function createEvent<T extends AgentCustomEvent["name"]>(
  name: T,
  value: EventValueType<T>,
): Extract<AgentCustomEvent, { name: T }> {
  return { name, value } as Extract<AgentCustomEvent, { name: T }>;
}
